### PR TITLE
Block unauthorized access to runtime actions, refactor authn/authz

### PR DESCRIPTION
**Pull request by slitviachenko**
## Description

- Block unauthorized access to runtime actions
- Refactor authn/authz

I added an optional isAuthorize parameter to manage authorization checks. If it doesn't make sense to require authorization for certain actions, we can disable it. By default, authorization is enabled for all web actions. I excluded authorization from the following actions:
- aem-genai-assistant/csv
- aem-genai-assistant/target

A service token is needed only for actions communicating with Firefall, so I created a new `asFirefallAction` decorator for that purpose.

Unfortunately, we cannot utilize AppBuilder's [built-in auth functionality](https://github.com/adobe/aem-genai-assistant/pull/282/commits/d8c3829e1b86bfc0e742f916f728945afd2403c6#diff-6491149c0e11feef942d4d9bd31d89fffaaa3ac4475b77c38443a5ea2d24a60aL21-L23) for user access token validation, as it works only within the scope of the org owning the application. It isn't suitable for cases where there is a need to share the application across different orgs. See more details [here](https://cq-dev.slack.com/archives/C013UDBFBGB/p1716438150500019).

## Related Issue
https://jira.corp.adobe.com/browse/SITES-21724

## How Has This Been Tested?
https://experience-qa.adobe.com/?devMode=true&shell_source=workspace&workspace=sites21724auth&shell_ims=prod#/aem/generate-variations/

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: actions/AuthAction.js [hunk 1]
```diff
@@ -142,44 +140,34 @@ async function checkForEarlyProductAccess(toggle, sdkKey, isInternal, org) {
   });
 }
 
-function asAuthAction(action) {
+function asAuthAction(action, isAuthorize = true) {
```
Comment (slitviachenko) [inline actions/AuthAction.js:143]: I added an optional `isAuthorize` parameter to manage authorization checks. If it doesn't make sense to require authorization for certain actions, we can disable it. By default, authorization is enabled for all web actions. I excluded authorization from the following actions:

- `aem-genai-assistant/csv`
- `aem-genai-assistant/target`

File: actions/AuthAction.js [hunk 2]
```diff
@@ -142,44 +140,34 @@ async function checkForEarlyProductAccess(toggle, sdkKey, isInternal, org) {
...
 
-      if (!await checkForEarlyProductAccess(earlyAccessToggle, ldSdkKey, isInternalUser, imsOrg)) {
-        throw new Error('Profile does not have access to the product');
+        if (!await checkForEarlyProductAccess(earlyAccessToggle, ldSdkKey, isInternalUser, imsOrg)) {
+          throw new Error('Profile does not have access to the product');
+        }
       }
     }
 
-    // If everything is okay, generate a service token
```
Comment (slitviachenko) [inline actions/AuthAction.js:178]: A service token is needed only for actions communicating with Firefall, so I created a new `asFirefallAction` decorator for that purpose.

File: actions/AuthAction.js [hunk 3]
```diff
@@ -77,6 +48,33 @@ async function isValidToken(endpoint, clientId, token) {
...
+  const clientId = params.IMS_CLIENT_ID;
+
+  // First, check if a bearer user access token is set
+  if (
+    params.__ow_headers
+    && params.__ow_headers.authorization
+    && params.__ow_headers.authorization.startsWith('Bearer ')
+  ) {
+    const accessToken = params.__ow_headers.authorization.substring('Bearer '.length);
+    // Validate the access token
```
Comment (slitviachenko) [inline actions/AuthAction.js:68]: Unfortunately, we cannot utilize AppBuilder's [built-in auth functionality](https://github.com/adobe/aem-genai-assistant/pull/282/commits/d8c3829e1b86bfc0e742f916f728945afd2403c6#diff-6491149c0e11feef942d4d9bd31d89fffaaa3ac4475b77c38443a5ea2d24a60aL21-L23) for user access token validation, as it works only within the scope of the org owning the application. It isn't suitable for cases where there is a need to share the application across different orgs. See more details [here](https://cq-dev.slack.com/archives/C013UDBFBGB/p1716438150500019).

File: actions/AuthAction.js [hunk 4]
```diff
@@ -142,44 +140,34 @@ async function checkForEarlyProductAccess(toggle, sdkKey, isInternal, org) {
...
-    // Check that the profile has access to the product
-    const imsProfile = await getImsProfile(imsEndpoint, clientId, accessToken);
-    if (!imsProfile) {
-      throw new Error('Failed to fetch profile');
-    }
+    if (isAuthorize) {
+      // Check that the profile has access to the product
+      const imsProfile = await getImsProfile(imsEndpoint, clientId, accessToken);
+      if (!imsProfile) {
+        throw new Error('Failed to fetch profile');
```
Comment (irenelagno) [inline actions/AuthAction.js:158]: You added a bunch of new error messages. Are we going to show them to the user through UI? if yes, we should add localization for these strings

Comment (slitviachenko) [reply actions/AuthAction.js:158]: @irenelagno , it looks like all error messages in this file are unlocalized, I see there is a ticket on our board to address that, https://jira.corp.adobe.com/browse/SITES-21934

File: actions/AuthAction.js (hunk 1 above)
Comment (lenaorobei) [inline actions/AuthAction.js:143]: To improve readability, I'd name the second param `isAuthorized` or `shouldAuthorize`

**Review by lenaorobei — APPROVED** [review]
Just a minor comment regarding naming

File: actions/AuthAction.js (hunk 1 above)
Comment (slitviachenko) [reply actions/AuthAction.js]: Renamed to `shouldAuthorize`, 👍

File: actions/AuthAction.js (hunk 4 above)
Comment (irenelagno) [reply actions/AuthAction.js:158]: @slitviachenko let's put comment that the new messages need to be localized as well

**Comment by irenelagno** [comment]
Generate image functionality doesn't work on PR.
Screen Shot 2024-05-30 at 10 47 10 AM

**Review by irenelagno — CHANGES_REQUESTED** [review]
Please check why generate images doesn't work on feature branch

**Comment by slitviachenko** [comment]
@irenelagno ,
>Please check why generate images doesn't work on feature branch

The issue was with the deployed code, I had the old Express URL value in my local `.env` file, `EXPRESS_SDK_URL=https://dev.cc-embed.adobe.com/sdk/1p/v4/4.1.14/CCEverywhere.js`, I changed it to `EXPRESS_SDK_URL=https://cc-embed.adobe.com/sdk/1p/v4/CCEverywhere.js`, and redeployed, now everything should work.

**Review by askayastha22 — APPROVED** [review]
LGTM

File: actions/AuthAction.js (hunk 4 above)
Comment (askayastha22) [reply actions/AuthAction.js:158]: Curious to understand the options that we have for localizing the text in runtime actions as we don't have the flexibility of React components.

Comment (askayastha22) [reply actions/AuthAction.js:158]: One solution may be to:
-  move the localizable text from Runtime actions to the appropriate localization file in the frontend.
- send a coded message like `error.profileFetchFailedMessage` from the Runtime actions for errors.
- figure out a way to call the `formatMessage` function correctly to display the correct localized text.

Comment (slitviachenko) [reply actions/AuthAction.js:158]: CC: @audrey-kho , ^^^ as you are working on https://jira.corp.adobe.com/browse/SITES-21934 , please take a look at @askayastha22 's suggestion

Comment (slitviachenko) [reply actions/AuthAction.js:158]: I would add to @askayastha22 's suggestion that we might return an error message and an error code. The message itself is for developers who can open Chrome tools for debugging, while the error code will be used for translation purposes on the frontend.
For example,
```
{
  "error": {
    "code": 10005,
    "message": "The access token is not valid"
  }
}
```

**Review by irenelagno — APPROVED** [review]

**Comment by vtsaplin** [comment]
You might want to consider splitting AuthAction into two parts: a generic AuthAction and one for exchanging the user IMS token for the service token. This removes unnecessary parameters and follows the "composition over configuration" principle.
We need to decide if we want to check the entitlement everywhere or just for certain actions. If it's only required for Firefall actions, you could combine the token exchange and entitlement check in the one authZ action.

**Comment by slitviachenko** [comment]
@vtsaplin , https://github.com/adobe/aem-genai-assistant/pull/287
Could you please take a look when you have time?

**Pull request merged**